src/Mutex.php

```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Mutex;
+
+/**
+ * The Mutex allows mutual execution of concurrent processes in order to prevent "race conditions".
+ *
+ * @see MutexInterface
+ */
+abstract class Mutex implements MutexInterface
```

Comment:
Problem was not solved. Factories still return `MutexInterface`, but not `Mutex`. If some driver will create mutex implementation that not extends `Mutex`, autorelease will not happen.

Reply:
This is a matter of implementing drivers. All drivers extend the abstract `Mutex` class, the user may need to write their own logic in the destructor. It will be more flexible this way.

Reply:
In this case strictness is more important than the flexibility. If user may need to write their own logic in the object shutdown, he may use `register_shutdown_function`. Moreover, we should not allow the interface to be implemented, mutex must be a final object without any interface that accepts releaser callback.

Reply:
I don't agree with you. If I need to write a standard implementation, I will extend 'Mutex'. If I need to do something specific, then I would like to have more options, I might not be satisfied with the final destructor with a non-mutable behavior.

If we give the user the opportunity to write their own drivers, why should we think what he can do there. This is a personal matter of the user. I think it is not right to limit its capabilities in this case.

Reply:
We should think about the consumers, not about the creators of drivers. Consumers should get stable drivers. Stability is achieved by restrictions.